workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/DialogCollator.kt

```diff
@@ -0,0 +1,214 @@
+ * Android notoriously doesn't give us any control over the z order of windows
+ * other than taking care to [show][android.app.Dialog.show] them in the right
+ * order, so keeping a pile of windows in sync with the shifting order of the
+ * defining list of [Overlay] rendering models is hard. [DialogCollator] manages
+ * that process. On each update it builds complete knowledge of the existing stack
+ * of windows, and so is able to decide if any instances need to be
+ *
+ * [LayeredDialogSessions.update] starts by calling [ViewEnvironment.establishDialogCollator],
+ * to register its existing list of [DialogSession], and provide a
+ * `MutableMap<String, KeyAndBundle>` that can be populated by the
+ * [classic view state][DialogSession.save] of any
+ */
+@WorkflowUiExperimentalApi
+internal class DialogCollator {
+  internal class SessionAndViewStateMap(
+    val id: UUID,
+    val sessions: List<DialogSession>
+  )
+
+  private class UpdateBlock(
+    val id: UUID,
+    val updates: List<DialogSessionUpdate>,
+    val callback: (List<DialogSession>) -> Unit
+  )
+
+  private val updateBlocks = mutableListOf<UpdateBlock>()
+
+  internal val establishedSessions = mutableListOf<SessionAndViewStateMap>()
+  internal var expectedUpdates = 0
+
+  /**
+   * Follow up call to [ViewEnvironment.establishDialogCollator]. Adds a set
+   * of update operations to be applied to the [DialogSession] set provided
+   * to that call. The updates are not applied until a matching call
+   * to [scheduleUpdates] has been received for each call to [establishDialogCollator].
+   *
+   * @param id unique runtime-only id used to pair calls to [establishDialogCollator]
+   * and [DialogCollator.scheduleUpdates]
+   *
+   * @param updates list of new [Overlay]s to show, each paired with a function
+   * to create or update a [DialogSession] to show it.
+   *
+   * @param callback called immediately after the given [updates] are applied.
+   * Provides the updated list of [DialogSession] that should replace those that
+   * were provided to [establishDialogCollator].
+   */
+  internal fun scheduleUpdates(
+    id: UUID,
+    updates: List<DialogSessionUpdate>,
+    callback: (List<DialogSession>) -> Unit
+  ) {
+    check(expectedUpdates > 0) {
+      "Each update() call must be preceded by a call to ViewEnvironment.establishDialogCollator, " +
+        "but expectedUpdates is $expectedUpdates"
+    }
+
+    this.updateBlocks.add(UpdateBlock(id, updates, callback))
+    if (--expectedUpdates == 0) doUpdate()
+  }
+
+  private fun getUpdatedModalIndex(): Int {
+    val overlays = updateBlocks.flatMap { it.updates.map { update -> update.overlay } }
+    return overlays.indexOfFirst { it is ModalOverlay }
+  }
+
+  private fun doUpdate() {
+    val visibleSessionsIterator = establishedSessions.asReversed().asSequence().flatMap {
+      it.sessions.map { session -> Pair(it.id, session) }
+    }.iterator()
+    val hiddenSessions = mutableListOf<Pair<UUID, DialogSession>>()
```

Comment:
I'll create a proper tuple class instead of using `Pair`.

Reply:
Also, `SessionAndViewStateMap` is a stale name left over from the first draft. I originally was carrying view state around and creating new dialog instances instead of just relying on `dimiss() / show()` to re-order existing ones.

Reply:
> I'll create a proper tuple class instead of using `Pair`.
Actually, since this pair is used only w/in this function, I'm concerned that
yet another tuple would be obfuscatory.

I've renamed and documented the collator's two core collection types:

```kotlin
  /** Sessions registered by a specific [LayeredDialogSessions] via [establishDialogCollator]. */
  internal class IdAndSessions(
    val id: UUID,
    val sessions: List<DialogSession>
  )

  /** Updates registered by a specific [LayeredDialogSessions] via [scheduleUpdates]. */
  private class ScheduledUpdates(
    val id: UUID,
    val updates: List<DialogSessionUpdate>,
    val callback: (List<DialogSession>) -> Unit
  )


  private val scheduledUpdates = mutableListOf<ScheduledUpdates>()

  internal val establishedSessions = mutableListOf<IdAndSessions>()
```

The list you asked about is used only within this method, to flatten the leftover bits of `establishedSessions` into a single `List<UUID + DialogSession>`. Maybe a comment to that effect woudl be enough?